IBX-1430: Revise REST API documentation for 4.x (#1591)
ezsystems/developer-documentation: docs/api/rest_api_usage.md

The response contains an `Allow` header, as specified in [chapter 14.7 of RFC 2616](https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.7), which lists the methods accepted by the resource.

```shell
curl -IX OPTIONS https://api.example.com/api/ibexa/v2/content/objects/1
```

```
OPTIONS /content/objects/1 HTTP/1.1
Host: api.example.com
```

```
HTTP/1.1 200 OK
Allow: PATCH,GET,DELETE,COPY
```

```shell
curl -IX OPTIONS https://api.example.com/api/ibexa/v2/content/locations/1/2
```

```
OPTIONS /content/locations/1/2 HTTP/1.1
Host: api.example.com
```

```
HTTP/1.1 200 OK
Allow: GET,PATCH,DELETE,COPY,MOVE,SWAP
```

### Request headers

There are HTTP headers to specify a REST request:

- [`Accept`](https://tools.ietf.org/html/rfc2616#section-14.1) describing the desired response type and format;
- [`Content-Type`](https://toos.ietf.org/html/rfc2616#section-14.17) describing the payload type and format;
- [`X-Siteaccess`](#siteaccess) specifying the target SiteAccess;
- `X-HTTP-Method-Override` allowing to pass a custom method while using `POST` method as previously seen in [HTTP method](#request-method).
- [`Destination`](#destination) specifying where to move an item

> <!-- - [`If-None-Match`](https://tools.ietf.org/html/rfc7232#section-3.2) reclaiming the cached response of a previously visited resource if still up-to-date using [HTTP Etag](https://tools.ietf.org/html/rfc7232#section-2.3). -->

Author: I have hidden HTTP Etag related entries has the feature doesn't seem usable yet. I'll remove those traces.

Examples of such requests are

- [copying a Content](rest_api_reference/rest_api_reference.html#managing-content-copy-content);
- [moving a Location and its subtree](rest_api_reference/rest_api_reference.html#managing-content-move-subtree)
- [swapping a Location with another](rest_api_reference/rest_api_reference.html#managing-content-swap-location)

### Request body

While some short scalar parameters can be passed in the URIs or as GET parameters, some resources needs heavier structured payloads passed in the request body, in particular the ones to create (`POST`) or update (`PATCH`) items.
In the [REST API reference](rest_api_reference/rest_api_reference.html), request payload examples are given when needed.

One example is the [creation of an authentication session](rest_api_authentication.md#establishing-a-session).

When creating a Content, the payload is particular if the ContentType has some [Image](field_types_reference/imagefield.md) or [BinaryFile](field_types_reference/binaryfilefield.md) fields as files need to be attached. See the example of a [script uploading images](#creating-content-with-binary-attachments) below.

> When searching for Contents (or Locations), the query grammar is also particular. See the [Search section](#search-view) below.

Author: ```suggestion
When searching for content (or location), the query grammar is also particular. See the [Search section](#search-view) below.
```
@DominikaK Same treatment for both, right?

> When searching for Contents (or Locations), the query grammar is also particular. See the [Search section](#search-view) below.

Reviewer: Actually, it's "content", "Content item", "Location", with this capitalization. Not quite consistent, I know, but let's stick to it until we agree on a better standard throughout.

> When searching for Contents (or Locations), the query grammar is also particular. See the [Search section](#search-view) below.

Author: So, maybe the following as in both cases you're looking for content but the object types of the result differ.
```suggestion
When searching for Content items (or Locations), the query grammar is also particular. See the [Search section](#search-view) below.
```

#### Creating content with binary attachments

The example below is a command-line script to upload images.

This script will

- receive an image path and optionally a name as command-line arguments,
- use the [HTTP basic authentication](rest_api_authentication.md#http-basic-authentication) assuming it is enabled,
- create a draft in the /Media/Images folder by `POST`ing data to [`/content/objects`](rest_api_reference/rest_api_reference.html#managing-content-create-content-item),

> - and, `PUBLISH` the draft through [`/content/objects/{contentId}/versions/{versionNo}`](rest_api_reference/rest_api_reference.html#managing-content-publish-a-content-version).

Author: ```suggestion
- and, publishes (`PUBLISH`) the draft through [`/content/objects/{contentId}/versions/{versionNo}`](rest_api_reference/rest_api_reference.html#managing-content-publish-a-content-version).
```

=== "XML"

    ``` php
    [[= include_file('code_samples/api/rest_api/create_image.xml.php', 0, None, '    ') =]]
    ```

=== "JSON"

    ``` php
    [[= include_file('code_samples/api/rest_api/create_image.json.php', 0, None, '    ') =]]
    ```

#### Search (`/view`)

The `/view` route allow to [search in the repository](../guide/search/search.md). It works close to its [PHP API counterpart](public_php_api_search.md).

The model allows combining criteria using the logical operators `AND`, `OR` and `NOT`.

Almost all [search criteria](../guide/search/criteria_reference/search_criteria_reference.md#search-criteria) are available on REST API. The suffix `Criterion` is added when used with REST API.

Almost all [sort clauses](../guide/search/sort_clause_reference/sort_clause_reference.md#sort-clauses) are available too. No prefix ou suffix for them.

> The search request HTTP header to type its body is `Content-Type: application/vnd.ibexa.api.ViewInput+xml` or `+json`.

Author: This header says what type of content is in the body. There is no such thing as a "body header". I'll try to found another way to present this.

The root node is `<ViewInput>` and it has two mandatory children: `<identifier>` and `<Query>`.

> `version=1.1` can be added to the `Content-Type` header to support the distinction between `ContentQuery` and `LocationQuery` instead of just `Query` which implicitly looked only for Contents.

Reviewer: ```suggestion
You can add `version=1.1` to the `Content-Type` header to support the distinction between `ContentQuery` and `LocationQuery` instead of just `Query` which implicitly looked only for Content items.
```

Author: ```suggestion
You can add `version=1.1` to the `Content-Type` header to support the distinction between `ContentQuery` and `LocationQuery` instead of just `Query` which implicitly looks only for Content items.
```

> The following examples will search for `article` and `news` typed Contents everywhere, search for Contents of all types directly under Location `123`, all those Contents must be in the `standard` Section.

Reviewer: ```suggestion
The following examples search for `article` and `news` typed Content items everywhere or for Content items of all types directly under Location `123`. All those Content items must be in the `standard` Section.
```

              },
              "SectionIdentifierCriterion": "standard"
            }
          },
          "limit": "10",
          "offset": "0",
          "SortClauses": { "ContentName": "ascending" }
        }
      }
    }
    ```

=== "JSON; 1.1"

    ```
    Content-Type: application/vnd.ibexa.api.ViewInput+json
    ```

    ``` json
    {
      "ViewInput": {
        "identifier": "test",
        "ContentQuery": {
          "Filter": {
            "AND": {
              "OR": {
                "ContentTypeIdentifierCriterion": [
                  "article",
                  "news"
                ],
                "ParentLocationIdCriterion": 123
              },
              "SectionIdentifierCriterion": "standard"
            }
          },
          "limit": "10",
          "offset": "0",
          "SortClauses": { "ContentName": "ascending" }
        }
      }
    }
    ```

!!! note

    In JSON, the structure for `ContentTypeIdentifierCriterion` with multiple values has a slightly different format as keys must be unique.
    In JSON, if there is only one `SortClauses`, it can be passed directly without an array to wrap it.

> Logical operators may be omitted. If criteria are of mixed types, they will be wrapped in an implicit `AND`. If they are of the same type, they will be wrapped in an implicit `OR`.

Reviewer: ```suggestion
You can omit logical operators. If Criteria are of mixed types, they are wrapped in an implicit `AND`.
If they are of the same type, they are wrapped in an implicit `OR`.
```

            </OR>
            <SectionIdentifierCriterion>standard</SectionIdentifierCriterion>
        </AND>
    </Filter>
    ```

=== "XML Implicit `AND`"

    ``` xml
    <Filter>
        <OR>
            <ContentTypeIdentifierCriterion>article</ContentTypeIdentifierCriterion>
            <ContentTypeIdentifierCriterion>news</ContentTypeIdentifierCriterion>
            <ParentLocationIdCriterion>123</ParentLocationIdCriterion>
        </OR>
        <SectionIdentifierCriterion>standard</SectionIdentifierCriterion>
    </Filter>
    ```

=== "JSON Explicit `AND`"

    ``` json
    "Filter": {
      "AND": {
        "OR": {
           "ContentTypeIdentifierCriterion": [
            "article",
            "news"
          ],
          "ParentLocationIdCriterion": 123
        },
        "SectionIdentifierCriterion": "standard"
      }
    },
    ```

=== "JSON Implicit `AND`"

    ``` json
    "Filter": {
      "OR": {
         "ContentTypeIdentifierCriterion": [
          "article",
          "news"
        ],
        "ParentLocationIdCriterion": 123
      },
      "SectionIdentifierCriterion": "standard"
    },
    ```

## Responses

### Response code

> The following list of available HTTP response status codes just give a quick hint of the meaning of a code. For code details per resource, see the [REST API reference](rest_api_reference/rest_api_reference.html).

Reviewer: ```suggestion
The following list of available HTTP response status codes gives an overview of the meaning of each code.
For code details per resource, see the [REST API reference](rest_api_reference/rest_api_reference.html).
```

| Code  | Message                | Description                                                                                                                                                                                                                                                  |
|-------|------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `200` | OK                     | The resource has been found.                                                                                                                                                                                                                                 |

> | `201` | Created                | The request to create a new item has succeeded; the response `Location` header indicates when the created item could be consulted.                                                                                                                           |

Reviewer: ```suggestion
| `201` | Created                | The request to create a new item has succeeded; the response `Location` header indicates where you can find the created item.                                                                                                                           |
```
?

> | `204` | No Content             | The request has succeeded and there is nothing to say about it in the response header nor body (for example when publishing or deleting).                                                                                                                    |

Reviewer: ```suggestion
| `204` | No Content             | The request has succeeded and there is no additional information in the response header or body (for example when publishing or deleting).                                                                                                                    |
```

| `301` | Moved Permanently      | The resource should not be accessed this way; the response `Location` header indicates the proper way.                                                                                                                                                       |
| `307` | Temporary Redirect     | The resource is available at another URL considered as its main; the response `Location` header indicates this main URL.                                                                                                                                     |
| `400` | Bad Request            | The input (payload) doesn't have the proper schema for the resource.                                                                                                                                                                                         |
| `401` | Unauthorized           | The user does not have the permission to make this request.                                                                                                                                                                                                  |

> | `403` | Forbidden              | The user has the permission but action can't be done because of repository logic (for example, when trying to create an item with an already existing ID or identifier, when attempting to update a version in another state than draft).                    |

Reviewer: ```suggestion
| `403` | Forbidden              | The user has the permission but action can't be performed because of Repository logic (for example, when trying to create an item with an already existing ID or identifier, when attempting to update a version in another state than draft).                    |
```

| `404` | Not Found              | The requested object (or a request data like the parent of a new item) hasn't been found.                                                                                                                                                                    |
| `405` | Method Not Allowed     | The requested resource does not support the HTTP verb that was used.                                                                                                                                                                                         |

> | `406` | Not Acceptable         | The request `accept` header is not supported.                                                                                                                                                                                                                |

Reviewer: ```suggestion
| `406` | Not Acceptable         | The request's `Accept` header is not supported.                                                                                                                                                                                                                |
```

| `409` | Conflict               | The request is in conflict with another part of the repository (for example, trying to create a new item with an identifier already used).                                                                                                                   |

> | `415` | Unsupported Media Type | The request payload media-type doesn't match the media-type specified in the request header                                                                                                                                                                  |

Reviewer: ```suggestion
| `415` | Unsupported Media Type | The request payload media type doesn't match the media type specified in the request header.                                                                                                                                                                  |
```

| `500` | Internal Server Error  | The server encountered an unexpected condition, usually an exception, which prevents it from fulfilling the request, like database down, permissions or configuration error.                                                                                 |
| `501` | Not Implemented        | Returned when the requested method has not yet been implemented. For [[= product_name =]], most of Users, User groups, Content items, Locations and Content Types have been implemented. Some of their methods, as well as other features, may return a 501. |

<!-- | `412` | Precondition Failed | The current ETag does not match with the one provided in the If-Match header | -->

### Response headers

> A resource's response may contain meta-data in its HTTP headers.

Reviewer: ```suggestion
A resource's response may contain metadata in its HTTP headers.
```

```

To possibly test it, just open a PHP shell in a terminal with `php -a` and copy-paste this code into it.

`$resource` URI should be edited to address the right domain.

On a freshly installed Ibexa DXP, `52` is the Content ID of the Front-Office home page. If necessary, substitute `52` with the Content ID of an item from your database.

A content creation example using PHP is available above in [Creating content with binary attachments section](#creating-content-with-binary-attachments)

### JS

One of the main reasons for this API is to help implement JavaScript / AJAX interaction. You can see here an example of an AJAX call that retrieves ContentInfo (e.g. metadata) for a Content item:

```javascript
var resource = '/api/ibexa/v2/content/objects/52',
    request = new XMLHttpRequest();

request.open('GET', resource, true);
request.setRequestHeader('Accept', 'application/vnd.ibexa.api.ContentInfo+json');
request.onload = function () {
    console.log(request.getAllResponseHeaders(), JSON.parse(request.responseText));
};
request.send();
```

To possibly test it, just copy-paste this code into your browser console alongside a page from your website (to share the domain).

> On a freshly installed Ibexa DXP, `52` is the Content ID of the Front-Office home page. If necessary, substitute `52` with the Content ID of an item from your database.

Author: ```suggestion
On a freshly installed Ibexa DXP, `52` is the Content ID of the home page. If necessary, substitute `52` with the Content ID of an item from your database.
```

- and, publishes (`PUBLISH`) the draft through [`/content/objects/{contentId}/versions/{versionNo}`](rest_api_reference/rest_api_reference.html#managing-content-publish-a-content-version).

=== "XML"

    ``` php
    [[= include_file('code_samples/api/rest_api/create_image.xml.php', 0, None, '    ') =]]
    ```

=== "JSON"

    ``` php
    [[= include_file('code_samples/api/rest_api/create_image.json.php', 0, None, '    ') =]]
    ```

#### Search (`/view`)

The `/view` route allows you to [search in the repository](../guide/search/search.md). It works similarly to its [PHP API counterpart](public_php_api_search.md).

The model allows combining criteria using the logical operators `AND`, `OR` and `NOT`.

Almost all [Search Criteria](../guide/search/criteria_reference/search_criteria_reference.md#search-criteria) are available in REST API. The suffix `Criterion` is added when used with REST API.

Almost all [Sort Clauses](../guide/search/sort_clause_reference/sort_clause_reference.md#sort-clauses) are available too. They require no additional prefix or suffix.

> The search request HTTP header to type its body is `Content-Type: application/vnd.ibexa.api.ViewInput+xml` or `+json`.

Reviewer: ```suggestion
The search request has a `Content-Type: application/vnd.ibexa.api.ViewInput+xml` or `+json` header to specify the format of its body's payload.
```
